### ISSUE TITLE
NF handling of previous() causes compilation errors

### ISSUE THREAD

**Issue by perost** [OpenModelica/OpenModelica#8297]
[Modelica.Clocked.Examples.Elementary.RealSignals.SampleWithADeffects](https://libraries.openmodelica.org/branches/master/Modelica_4.0.0/files/Modelica_4.0.0_Modelica.Clocked.Examples.Elementary.RealSignals.SampleWithADeffects.err) (among others) currently fail during C compilation with the NF. This is because the NF flattens `previous(seedState)` to `previous(seedState)` while the OF flattens it to `{previous(seedState[1]), previous(seedState[2]), previous(seedState[3])}`

The [specification](https://specification.modelica.org/master/synchronous-language-elements.html#clocked-state-variables) just says that `previous` takes a component expression or a parameter expression (is that correct?) and that the return type is the same as the input type. The NF does not check those restrictions at the moment and just allowed anything as argument to `previous`, but that's relatively easy to add.

The question is though, should the NF also expand `previous` in the same way that the OF does or do we want to improve the backend so it can handle `previous` of array variables too?

**Comment by AnHeuermann** [comment]
>The specification just says that previous takes a component expression or a parameter expression (is that correct?) and that the return type is the same as the input type. 

I understand it in that way. So adding a check would be great.

> The question is though, should the NF also expand previous in the same way that the OF does or do we want to improve the backend so it can handle previous of array variables too?

I'm currently trying to improve the C runtime to work correctly with the current status of the backend (expanded variables). But using array variables shouldn't really be that much of a difference for the runtime.

So I'm for improving the Backend to work with non-expanded variables. But maybe after https://github.com/OpenModelica/OpenModelica/issues/7854 is mostly finished.

**Comment by perost** [comment]
> > The specification just says that previous takes a component expression or a parameter expression (is that correct?) and that the return type is the same as the input type.
> 
> I understand it in that way. So adding a check would be great.

Ok, so the argument can be any expression with at most parameter variability? E.g. `previous(17)` is fine? It doesn't matter to me, I can easily check that, just seems weird.

But if I add checking for that I may as well expand the calls too and get those models working in 1.19, then we can disable it later if the backend is improved.

**Comment by AnHeuermann** [comment]
> Ok, so the argument can be any expression with at most parameter variability? E.g. previous(17) is fine? It doesn't matter to me, I can easily check that, just seems weird.

It has to be a component expression [16.2.3 Argument Restrictions (Component Expression)](https://specification.modelica.org/master/synchronous-language-elements.html#def:component-expression) . But if `constant Real n=17` is a component expression (?) then the standard is saying that `previous(17)` should be legal.
Will it always be `17` then? Not really useful.

**Comment by perost** [comment]
> > Ok, so the argument can be any expression with at most parameter variability? E.g. previous(17) is fine? It doesn't matter to me, I can easily check that, just seems weird.
> 
> It has to be a component expression [16.2.3 Argument Restrictions (Component Expression)](https://specification.modelica.org/master/synchronous-language-elements.html#def:component-expression) . But if `constant Real n=17` is a component expression (?) then the standard is saying that `previous(17)` should be legal. Will it always be `17` then? Not really useful.

It's not a component expression, but it is a [parameter expression](https://specification.modelica.org/master/operators-and-expressions.html#parameter-expressions). That's the part I'm confused about, because I can't see any examples in the specification where that's actually used.

**Comment by AnHeuermann** [comment]
Ah I see. We should probably open a ticket about that in the Modelica Specification.

**Comment by perost** [comment]
#8301 adds some error checking for `previous` and also expands it (unless `-d=-nfScalarize` is used, not sure if that actually works but I'm not going to mess with it).

To disable the expansion, just remove the call to `ExpandExp.expand` in `NFBuiltinCall.typePreviousCall`.

Should I close this once my PR is merged, or do we want to keep this open as a reminder to improve the backend too?

**Comment by casella** [comment]
> The [specification](https://specification.modelica.org/master/synchronous-language-elements.html#clocked-state-variables) just says that `previous` takes a component expression or a parameter expression (is that correct?)

That is actually weird. I opened [#3074](https://github.com/modelica/ModelicaSpecification/issues/3074) on the Modelica Specification issue tracker.

> The question is though, should the NF also expand `previous` in the same way that the OF does or do we want to improve the backend so it can handle `previous` of array variables too?

I think so. If you have an array of models `m[:]` containing something like `y = previous(y) + 1`, flattening it would produce `m.y = previous(m.y) + 1`, where `m.y` is an array, and of course we need to support that.

**Comment by perost** [comment]
My change to expand `previous` caused as many issues as it solved, because it turns out it's not always possible to expand it. In #8323 I've now instead implemented custom polymorphic types, so we can just define `previous` in ModelicaBuiltin as:
```Modelica
function previous<__Scalar>
  input __Scalar u;
  output __Scalar y;
  external "builtin";
end previous;
```
and then we have a rule in the type checking to treat `__Scalar` as any scalar type. This means we don't need any custom typing of `previous`, and it will just be typed and vectorized like any other function call.

**Comment by casella** [comment]
Excellent! I noticed the regressions, I hope they are fixed by #8323, let's see the next library test report.

**Comment by perost** [comment]
> Excellent! I noticed the regressions, I hope they are fixed by #8323, let's see the next library test report.

The (https://libraries.openmodelica.org/branches/history/master/2021-12-14%2014:43:33..2021-12-15%2012:06:48.html)[report] says the regressions are now fixed and the improvements still remain, so I think we can close this. If @AnHeuermann feel that we need a reminder for improving the backend he can open a more appropriate issue for that.
